From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: mrhua269 <mrhua269@gmail.com>
Date: Tue, 28 Oct 2025 23:19:33 +0800
Subject: [PATCH] Fix possible delayed POI updates while block updating

For the on-region block updates, we would run the poi manager add or remove immediately instead of scheduling to target as we could 100% make sure it is running correctly and those unnecessary scheduling would cause delay between block update and poi update which makes multiple portal being created in a single area

diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 87377bba2ebdc0eba5b7b212d971994c35ca0b62..b9b5e3dc5aa6a5502fb0d9c5ac61fed8183f9cb0 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -2394,11 +2394,13 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
         Optional<Holder<PoiType>> optional1 = PoiTypes.forState(newState);
         if (!Objects.equals(optional, optional1)) {
             BlockPos blockPos = pos.immutable();
-            optional.ifPresent(holder -> io.papermc.paper.threadedregions.RegionizedServer.getInstance().taskQueue.queueChunkTask(this, blockPos.getX() >> 4, blockPos.getZ() >> 4, () -> { // Folia - region threading
+            final boolean isOnTargetRegion = ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(this, blockPos.getX() >> 4, blockPos.getZ() >> 4); // Luminol - Fix possible delayed POI updates while block updating
+            final java.util.concurrent.Executor targetRegion = isOnTargetRegion ? Runnable::run : task -> io.papermc.paper.threadedregions.RegionizedServer.getInstance().taskQueue.queueChunkTask(this, blockPos.getX() >> 4, blockPos.getZ() >> 4, task);  // Luminol - Fix possible delayed POI updates while block updating
+            optional.ifPresent(holder -> /*io.papermc.paper.threadedregions.RegionizedServer.getInstance().taskQueue.queueChunkTask(this, blockPos.getX() >> 4, blockPos.getZ() >> 4,*/targetRegion.execute( () -> { // Folia - region threading  // Luminol - Fix possible delayed POI updates while block updating
                 this.getPoiManager().remove(blockPos);
                 DebugPackets.sendPoiRemovedPacket(this, blockPos);
             }));
-            optional1.ifPresent(holder -> io.papermc.paper.threadedregions.RegionizedServer.getInstance().taskQueue.queueChunkTask(this, blockPos.getX() >> 4, blockPos.getZ() >> 4, () -> { // Folia - region threading
+            optional1.ifPresent(holder -> /*io.papermc.paper.threadedregions.RegionizedServer.getInstance().taskQueue.queueChunkTask(this, blockPos.getX() >> 4, blockPos.getZ() >> 4,*/targetRegion.execute( () -> { // Folia - region threading  // Luminol - Fix possible delayed POI updates while block update // Luminol - Fix possible delayed POI updates while block updating
                 // Paper start - Remove stale POIs
                 if (optional.isEmpty() && this.getPoiManager().exists(blockPos, ignored -> true)) {
                     this.getPoiManager().remove(blockPos);
